js/src/forum/states/SearchState.js

```diff
@@ -0,0 +1,100 @@
+export default class SearchState {
+  constructor(cachedSearches = [], searchRoute = 'index') {
+    this.searchRoute = searchRoute;
+
+    this.cachedSearches = cachedSearches;
+  }
+
+  getValue() {
+    if (this.value === undefined) {
+      this.value = this.getCurrentSearch() || '';
+    }
+
+    return this.value;
+  }
+
+  setValue(value) {
+    this.value = value;
+  }
+
+  /**
+   * Get URL parameters that stick between filter changes.
+   *
+   * @return {Object}
+   */
+  stickyParams() {
+    return {
+      sort: m.route.param('sort'),
+      q: m.route.param('q'),
+    };
+  }
+
+  /**
+   * Get parameters to pass to the DiscussionList component.
+   *
+   * @return {Object}
+   */
+  params() {
+    const params = this.stickyParams();
+
+    params.filter = m.route.param('filter');
+
+    return params;
+  }
+
+  /**
+   * Return the current search query, if any. This is implemented to activate
+   * the search box in the header.
+   *
+   * @see Search
+   * @return {String}
+   */
+  getCurrentSearch() {
+    return this.params().q;
```

Comment:
Previous call sites checked for the controller supporting search, right? So we have a slight change in behavior - in theory, other places having a `q` query param would now be interpreted as having a current search, too. Not sure if this is a practical problem. :thinking:

Reply:
Now that value is stored in SearchState, we could always just return that?

Reply:
That won't work - `getCurrentSearch()` is the fallback value called in `getValue()`, so you'd have an infinite loop there.

That said, what about my question? We don't check whether the current page / controller supports search anymore, right?

Reply:
> I do not like the search PR
> This code is more opaque than tar
> This jumbled madness breaks my mind,
> and while an answer we shall find,
> what that shall be, I can't forsee
> I do not like this, Franz Liedke.

But on a serious note, we don't but we could just move that check here and return null if the page isn't searchable?

Reply:
I'll be honest, I don't really like the naming here. I think it needs to be clearer to indicate that getCurrentSearch is whatever was searched on the last page refresh, and getValue is the current value of the search input. I feel like there's way too many methods involved in search and it could probably be clearer. Not sure where to go with it though.